esx_service/vmdk_ops.py

```diff
@@ -1670,22 +1681,43 @@ def set_vol_opts(name, tenant_name, options):
 
     return False
 
-def msg_about_signal(signalnum, details="exiting"):
-    logging.warn("Received signal num: %d, %s.", signalnum, details)
-    logging.warn("Operations in flight will be terminated")
+def wait_ops_in_flight():
+    # Wait for the event indicating all in-flight ops are drained
+    eventReceived = opsCounter.wait(WAIT_OPS_TIMEOUT)
+    if (eventReceived):
+        logging.info("All in-flight operations are completed. Exiting...")
+        sys.exit(0)
```

Comment:
sys.exit ony works on the main thread. I believe I mentioned it in some other comments. You may need to use os.signal(os.getpid(), signal.SIGKILL)

Reply:
I've replied in previous comments - copy it here:

"You are right about the behavior of sys.exit(), but the current solution does not really rely on this behavior. We spawned a thread to wait for ops in flight, and also break/exit the main loop when barrier is set. So it still work as expected."

Do you still think we need to use os.signal(os.getpid(), signal.SIGKILL)?

Reply:
Ah, sorry I must have missed your reply (high traffic !) .

Ah, I see what's going on. The waiting thread does not really forces the service to exit, thus the main thread will wait on get_one_op forever, or until there is a request from the client, and thus forces this strange code in the main loop to avoid the block. (there is another issue of NOT returning from get_one_op on client request, but I think it's unlikely since it's basic functionality). 

So, this is how it should work: 
- waiting thread forces the service to quit when the wait is done. This will interrupt whatever main thread is doing
- main thread does not have to have any special handling of the break on wait complete. It is not it's responsibility. Instead, it just rejects operations when the barrier is UP
- if you see the ops not getting through when the barrier is up, let's investigate offline. 

this should make the overall flow much simpler and clearer

Reply:
I will push one more commit, which I feel more comfortable with. Let's discuss based on the new commit.